### ISSUE TITLE
Error messages for type application mismatch are not clear

### ISSUE THREAD

**Issue by bugarela** [informalsystems/quint#1361]
```
/home/gabriela/projects/quint/local/test.qnt:2:13 - error: [QNT000] Couldn't unify set and list
Trying to unify Set[_t3] and List[int]
Trying to unify (Set[_t3], (_t3) => _t4) => Set[_t4] and (List[int], (int) => int) => _t5

2:   val foo = [1, 2].map(x => x + 1)
               ^^^^^^^^^^^^^^^^^^^^^^

error: typechecking failed
```

Could be something like:
```
/home/gabriela/projects/quint/local/test.qnt:2:13 - error: [QNT000] Operator `map` expected `Set[_t3]` as its first argument, but the provided argument `[1, 2]` is a `List[int]`.
  Expected: Set[_t3]
  Got: List[int]

Type signature for `map` is (Set[a], (a) => b) => Set[b]

2:   val foo = [1, 2].map(x => x + 1)
               ^^^^^^^^^^^^^^^^^^^^^^

error: typechecking failed
```

**Comment by kalidatony** [comment]
Hey @bugarela, is this task available? Can I work on this?

**Comment by bugarela** [comment]
@kalidatony of course! I do warn you that this can be on the harder side, but feel free to try it as any advancements will be super helpful. Don't hesitate to ping me if you need some assistance 😄

**Comment by kalidatony** [comment]
Hi @bugarela,

Thank you for the response. Before starting, I would like to share my plans and get your opinion.

My plan is roughly:
1.  **Enhance Unification Context:** Modify the unification process (likely within `unify` in `constraintSolver.ts`) to capture more specific details when a type mismatch occurs in the context of an operator/function application. This would include the operator name, the specific argument, expected type, actual type, and the operator's signature.
2.  **Structured Error Data:** Introduce a more structured way to hold this information, perhaps by augmenting the existing `ErrorTree` or introducing a new dedicated structure for these kinds of errors. This would make the information more robustly available to the formatting stage.
3.  **Improved Formatting:** Update the error message formatting logic (which seems to involve `errorTreeToString`) to use this new structured data to produce the clearer messages suggested in the issue.
4.  **Testing:** Add new test cases to verify the improved messages for specific scenarios like the `[1, 2].map(x => x + 1)` example.

I'm considering a couple of ways to approach step 2 (Structured Error Data):
    a) Incrementally adding optional fields to the current `ErrorTree`.
    b) Defining a new, more strongly-typed structure for these specific error details, which `ErrorTree` could then optionally hold. I'm leaning towards (b) for better long-term maintainability and extensibility, even if it's a bit more work upfront.


*   Does this general direction make sense?
*   Are there any particular parts of the codebase you'd recommend I pay special attention to or any potential pitfalls I should be aware of for this kind of change?
*   Is the idea of a more structured error object (rather than just adding fields to `ErrorTree`) something you'd be open to?

Any initial guidance would be super helpful!

Thanks!

**Comment by bugarela** [comment]
Hi @kalidatony. The direction sounds good - I also feel like we need to improve `ErrorTree` and that it is not currently enough to hold all the data we need to report meaningful errors. I haven't thought much about what this structure should look like, but your ideas sound good to me.

About pitfalls, one I can think of is it can be hard to combine errors when there are two of them (i.e. in the same operator application). There's a need to decide when something is part of the same error and when we need to report two separate errors to the user.

**Comment by kalidatony** [comment]
@bugarela I have send a PR #1668 . Please give it a look when you are free.

**Comment by kalidatony** [comment]
@bugarela I have send a PR #1668 . Please give it a look when you are free.
